go-controller/pkg/ovn/namespace.go

```diff
@@ -3,10 +3,42 @@ package ovn
 import (
 	"github.com/Sirupsen/logrus"
 	kapi "k8s.io/api/core/v1"
+	"os/exec"
+	"strings"
 	"sync"
 	"time"
 )
 
+func (oc *Controller) syncNamespaces(namespaces []interface{}) {
+	expectedNs := make(map[string]bool)
+	for _, nsInterface := range namespaces {
+		ns, ok := nsInterface.(*kapi.Namespace)
+		if !ok {
+			logrus.Errorf("Spurious object in syncNamespaces: %v", nsInterface)
+			continue
+		}
+		expectedNs[ns.Name] = true
+	}
+	// get the list of address sets from OVN
+	output, err := exec.Command(OvnNbctl, "--data=bare", "--no-heading",
```

Comment:
For this bit, and the same code in syncNetworkPolicies(), I was going to suggest making a discrete function for enumerating the address_sets, something like:

```golang
func forEachAddressSet(iteratorFn func(string, string, string)) error {
    output, err := exec.Command(OvnNbctl, "--data=bare", "--no-heading",
        "--columns=external_ids", "find", "address_set").Output()
    if err != nil {
        logrus.Errorf("Error in obtaining list of address sets from OVN: %v", err)
        return
    }

    for _, addrSet := range strings.Fields(string(output)) {
        if !strings.HasPrefix(addrSet, "name=") {
            continue
        }
        addrSetName := addrSet[5:]
        names := strings.Split(addrSetName, ".")
        asNamespace := names[0]
        asName := ""
        if len(names) >= 2 {
            asName = names[1]
        }
        iteratorFn(addrSetName, asNamespace, asName)
    }
    return nil
}

    err := forEachAddressSet(func(addrSetName, asNamespace, asName string) error {
        if asName == "" && !isDefaultKubeNamespace(asNamespace) && !expectedNs[asNamespace] {
            // delete the address sets for this policy from OVN
           oc.deleteAddressSet(hashedAddressSet(addrSetName))
        }
    })

    err := forEachAddressSet(func(addrSetName, asNamespace, asName string) error {
        if asName != "" && !expectedPolicies[asNamespace][asName] {
            // policy doesn't exist on k8s delete acl rules from OVN
            oc.deleteAclsPolicy(asNamespace, asName)
            // delete the address sets for this policy from OVN
            oc.deleteAddressSet(hashedAddressSet(addrSetName))
    }
```

Or something like that.  It would consolidate the address_set parsing code and simplify each location that uses addres sets.

Reply:
Nice and clean. Thanks Dan!